tracer/src/Datadog.Trace/AppSec/Security.cs

```diff
@@ -243,30 +216,152 @@ private static void AddAppsecSpecificInstrumentations()
                 var payload = InstrumentationDefinitions.GetDerivedDefinitions(InstrumentationCategory.AppSec);
                 NativeMethods.InitializeProfiler(payload.DefinitionsId, payload.Definitions);
-                foreach (var def in payload.Definitions)
-                {
-                    def.Dispose();
-                }
+                derived = payload.Definitions.Length;
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, ex.Message);
+            }
+
+            Log.Information($"{defs} AppSec definitions and {derived} derived definitions added to the profiler.");
+        }
 
-                Log.Information<int>("The profiler has been initialized with {count} AppSec derived definitions.", payload.Definitions.Length);
+        private static void RemoveAppsecSpecificInstrumentations()
+        {
+            int defs = 0, derived = 0;
+            try
+            {
+                Log.Debug("Adding CallTarget AppSec integration definitions to native library.");
+                var payload = InstrumentationDefinitions.GetAllDefinitions(InstrumentationCategory.AppSec);
+                NativeMethods.RemoveCallTargetDefinitions(payload.DefinitionsId, payload.Definitions);
+                defs = payload.Definitions.Length;
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, ex.Message);
+            }
+
+            try
+            {
+                Log.Debug("Adding CallTarget appsec derived integration definitions to native library.");
+                var payload = InstrumentationDefinitions.GetDerivedDefinitions(InstrumentationCategory.AppSec);
+                NativeMethods.RemoveCallTargetDefinitions(payload.DefinitionsId, payload.Definitions);
+                derived = payload.Definitions.Length;
             }
             catch (Exception ex)
             {
                 Log.Error(ex, ex.Message);
             }
+
+            Log.Information($"{defs} AppSec definitions and {derived} derived definitions removed from the profiler.");
         }
 
-        /// <summary>
-        /// Frees resources
-        /// </summary>
+        /// <summary> Frees resources </summary>
         public void Dispose()
         {
             _waf?.Dispose();
         }
 
+        private void FeaturesProductConfigChanged(object sender, ProductConfigChangedEventArgs e)
+        {
+            var features = e.GetDeserializedConfigurations<Features>().FirstOrDefault();
+            if (features != null)
+            {
+                _settings.Enabled = features.Asm.Enabled;
+                UpdateStatus();
+            }
+        }
+
+        private void UpdateStatus()
+        {
+            if (_enabled == _settings.Enabled) { return; }
+            lock (_settings)
+            {
+                if (_settings.Enabled)
+                {
+                    if (_waf != null)
+                    {
+                        _waf.Dispose();
```

Comment:
I guess we arrive here if there's a change in remote config? here don't we wanna call again `DisableWaf` on top of the `Dispose`? otherwise won't it re-add the instrumentations that werent disposed before? if they're only disposed at shutdown..  🤔 just thinkin out loud^^

Reply:
`EnableWaf` is covered with `if (!_enabled)`. You can't call it twice in a row. The events and the instrumentation use the _waf, so there is no lost event neither double initialization.